api/src/services/eml-service.ts

```diff
@@ -0,0 +1,224 @@
+
+    const organizationName =
+      response.find((item) => item.constant_name === 'ORGANIZATION_NAME_FULL')?.character_value || '';
+    const onlineUrl = response.find((item) => item.constant_name === 'ORGANIZATION_URL')?.character_value || '';
+
+    return { organizationName: organizationName, onlineUrl: onlineUrl };
+  }
+
+  /**
+   * Decorates an EML object, adding additional BioHub metadata to the original EML.
+   *
+   * @param {Record<string, any>} eml
+   * @return {*}  {Promise<Record<string, any>>}
+   * @memberof EMLService
+   */
+  async decorateEML(eml: Record<string, any>): Promise<Record<string, any>> {
+    // The config for a list of patch operations to perform on the EML.
+    // Note: the order of these patches may be important. Below, when patches are being applied, they are applied in
+    // order, allows iterative patches to be supported.
+    const patches = [
+      {
+        // Set the `@_system` attribute on the `eml:eml` node.
+        path: '$..eml:eml',
+        property: '@_system',
+        value: this.getSystemURL()
+      },
+      {
+        // Set the `@_system` attribute on the `dataset` node.
+        path: '$..eml:eml.dataset',
+        property: '@_system',
+        value: this.getDatasetSystemURL()
+      },
+      {
+        // Set the `@_system` attribute on the `project` node.
+        path: '$..eml:eml.dataset.project',
+        property: '@_system',
+        value: this.getDatasetSystemURL()
+      },
+      {
+        // Set the `@_system` attribute on all `relatedProject` nodes.
+        path: '$..eml:eml..relatedProject[*]',
+        property: '@_system',
+        value: this.getDatasetSystemURL()
+      },
+      {
+        // Set the `@_provider` attribute on all `taxonId` nodes that are a descendent of `taxonomicCoverage`
+        path: '$..eml:eml..taxonomicCoverage..taxonId',
+        property: '@_provider',
+        value: this.getTaxonomicProviderURL()
+      },
+      {
+        // Add a `metadataProvider` node if one does not already exist
+        path: '$..eml:eml.dataset',
+        property: 'metadataProvider',
+        value: []
+      },
+      {
+        // Append an object to the `metadataProvider` node.
+        path: '$..eml:eml.dataset.metadataProvider',
+        property: '-', // append to the end of the array
+        value: await this.getMetadataProviderNode()
+      }
+    ];
+
+    const patchOperations: Operation[] = [];
+
+    // Iterates over `patches` in order from first to last, applying all resulting patch operations one by one.
+    // Note: subsequent patches will operate on the output of the previous patches. This means iterative patches are
+    // possible, if needed.
+    for (const patch of patches) {
+      try {
+        const paths = JSONPath({
+          path: patch.path,
+          json: eml,
+          resultType: 'all'
+        });
+
+        paths.forEach(async (path: any) => {
```

Comment:
Is this callback asynchronous?

Reply:
Nope! I think I did have an async call at one point and removed it.

Fixed.